pkg/apis/openstack/validation/infrastructure.go

```diff
@@ -74,66 +77,71 @@ func ValidateInfrastructureConfigUpdate(oldConfig, newConfig *api.Infrastructure
+func validateFloatingPoolNameConstraints(fps []api.FloatingPool, domain, region string, name string, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	_, errs := findFloatingPool(fps, region, name, fldPath.Child("floatingPoolName"))
+	_, errs := findFloatingPool(fps, domain, region, name, fldPath.Child("floatingPoolName"))
 	allErrs = append(allErrs, errs...)
 	return allErrs
 }
 
-func findFloatingPool(floatingPools []api.FloatingPool, shootRegion, floatingPoolName string, fldPath *field.Path) (*api.FloatingPool, field.ErrorList) {
+func findFloatingPool(floatingPools []api.FloatingPool, domain, shootRegion, floatingPoolName string, fldPath *field.Path) (*api.FloatingPool, field.ErrorList) {
 	var (
-		allErrs                = field.ErrorList{}
-		validValues            []string
-		globalCandidate        *api.FloatingPool
-		globalCandidateScore   int
-		regionalCandidate      *api.FloatingPool
-		regionalCandidateScore int
-		regionalFound          bool
+		allErrs     = field.ErrorList{}
+		validValues = map[string]struct{}{}
 	)
 
-	for _, fp := range floatingPools {
-		// store the first non-regional image for fallback value if no floating pool for the given
-		// region was found
-		if fp.Region == nil && !regionalFound {
-			if match, score := utils.SimpleMatch(fp.Name, floatingPoolName); match {
-				if globalCandidate == nil || globalCandidateScore < score {
-					f := fp
-					globalCandidate = &f
-					globalCandidateScore = score
-				}
-			}
-			continue
-		}
+	found := findFloatingPoolSub(floatingPools, &domain, &shootRegion, floatingPoolName, validValues)
+	if found != nil {
+		return found, allErrs
+	}
+	found = findFloatingPoolSub(floatingPools, nil, &shootRegion, floatingPoolName, validValues)
+	if found != nil {
+		return found, allErrs
+	}
+	found = findFloatingPoolSub(floatingPools, &domain, nil, floatingPoolName, validValues)
+	if found != nil {
+		return found, allErrs
+	}
+	found = findFloatingPoolSub(floatingPools, nil, nil, floatingPoolName, validValues)
+	if found != nil {
+		return found, allErrs
+	}
 
-		// floating pool for the given region found
-		if fp.Region != nil && *fp.Region == shootRegion {
-			regionalFound = true
-			validValues = append(validValues, fp.Name)
+	validValuesList := []string{}
+	for key := range validValues {
+		validValuesList = append(validValuesList, key)
+	}
+	sort.Strings(validValuesList)
+	allErrs = append(allErrs, field.NotSupported(fldPath, floatingPoolName, validValuesList))
+	return nil, allErrs
+}
+
+func findFloatingPoolSub(floatingPools []api.FloatingPool, domain, shootRegion *string, floatingPoolName string, validValues map[string]struct{}) *api.FloatingPool {
+	var (
+		candidate      *api.FloatingPool
+		candidateScore int
+	)
+
+	for _, fp := range floatingPools {
+		if reflect.DeepEqual(domain, fp.Domain) && reflect.DeepEqual(shootRegion, fp.Region) {
```

Comment:
This introduces a change to the general validation logic. Previously, we forbade to use a non-regional FloatingPool if the CloudProfile specified FloatingPool(s) for the region used by the shoot.
A regional configuration has not been part of FloatingPools since day one. Hence, the assumption is that a regional setting for FloatingPools overrules a non regional setting.

I'm not sure if we proactively want to open this restriction. @rfranzke WDYT?

Reply:
I have added a table test with all  relevant cases to make reasoning easier. See pkg/apis/openstack/validation/infrastructure_test.go `Describe("#FindFloatingPool")...`

Reply:
Okay, but what is the motivation to open up the mentioned restriction?

Reply:
With the last push the logic has been reverted to the original behaviour. Regional or domain settings overrule any non-regional setting.